File: common/src/main/java/juuxel/adorn/compat/rei/client/AdornReiClient.java
package juuxel.adorn.compat.rei.client;

import juuxel.adorn.block.AdornBlocks;
import juuxel.adorn.compat.rei.AdornReiServer;
import juuxel.adorn.lib.AdornTags;
import me.shedaniel.rei.api.client.plugins.REIClientPlugin;
import me.shedaniel.rei.api.client.registry.category.CategoryRegistry;
import me.shedaniel.rei.api.client.registry.entry.CollapsibleEntryRegistry;
import me.shedaniel.rei.api.client.registry.screen.ScreenRegistry;
import me.shedaniel.rei.api.common.util.EntryIngredients;
import me.shedaniel.rei.api.common.util.EntryStacks;
import net.minecraft.world.item.Item;
import net.minecraft.tags.TagKey;
import net.minecraft.network.chat.Component;
import net.minecraft.util.Util;

public class AdornReiClient implements REIClientPlugin {
    @Override
    public void registerCategories(CategoryRegistry registry) {
        registry.add(new BrewerCategory());
        registry.addWorkstations(AdornReiServer.BREWER, EntryStacks.of(AdornBlocks.BREWER.get()));
    }

    @Override
    public void registerCollapsibleEntries(CollapsibleEntryRegistry registry) {
        add(registry, AdornTags.SOFAS.item());
        add(registry, AdornTags.CHAIRS.item());
        add(registry, AdornTags.TABLES.item());
        add(registry, AdornTags.DRAWERS.item());
        add(registry, AdornTags.KITCHEN_COUNTERS.item());
        add(registry, AdornTags.KITCHEN_CUPBOARDS.item());
        add(registry, AdornTags.KITCHEN_SINKS.item());
        add(registry, AdornTags.POSTS.item());
        add(registry, AdornTags.PLATFORMS.item());
        add(registry, AdornTags.STEPS.item());
        add(registry, AdornTags.SHELVES.item());
        add(registry, AdornTags.CHIMNEYS.item());
        add(registry, AdornTags.COFFEE_TABLES.item());
        add(registry, AdornTags.BENCHES.item());
        add(registry, AdornTags.CRATES.item());
        add(registry, AdornTags.TABLE_LAMPS.item());
        add(registry, AdornTags.CANDLELIT_LANTERNS.item());
        add(registry, AdornTags.COPPER_PIPES.item());
        add(registry, AdornTags.PAINTED_PLANKS.item());
        add(registry, AdornTags.PAINTED_WOOD_SLABS.item());
        add(registry, AdornTags.PAINTED_WOOD_STAIRS.item());
        add(registry, AdornTags.PAINTED_WOOD_FENCES.item());
        add(registry, AdornTags.PAINTED_WOOD_FENCE_GATES.item());
        add(registry, AdornTags.PAINTED_WOOD_PRESSURE_PLATES.item());
        add(registry, AdornTags.PAINTED_WOOD_BUTTONS.item());
    }

    private static void add(CollapsibleEntryRegistry registry, TagKey<Item> tag) {
        // matches the translation keys used by EMI as well
        var name = Component.translatable(Util.makeDescriptionId("tag.item", tag.location()));
        registry.group(tag.location(), name, EntryIngredients.ofItemTag(tag));
    }

    @Override
    public void registerScreens(ScreenRegistry registry) {
        registry.registerDraggableStackVisitor(new TradingStationDraggableStackVisitor());
    }
}

File: common/src/main/kotlin/juuxel/adorn/compat/rei/client/AdornReiClient.kt
package juuxel.adorn.compat.rei.client

import juuxel.adorn.block.AdornBlocks
import juuxel.adorn.compat.rei.AdornReiServer
import juuxel.adorn.compat.rei.BrewerDisplay
import juuxel.adorn.lib.AdornTags
import juuxel.adorn.recipe.AdornRecipes
import juuxel.adorn.recipe.FluidBrewingRecipe
import juuxel.adorn.recipe.ItemBrewingRecipe
import me.shedaniel.rei.api.client.plugins.REIClientPlugin
import me.shedaniel.rei.api.client.registry.category.CategoryRegistry
import me.shedaniel.rei.api.client.registry.display.DisplayRegistry
import me.shedaniel.rei.api.client.registry.entry.CollapsibleEntryRegistry
import me.shedaniel.rei.api.client.registry.screen.ScreenRegistry
import me.shedaniel.rei.api.common.util.EntryIngredients
import me.shedaniel.rei.api.common.util.EntryStacks
import net.minecraft.item.Item
import net.minecraft.tag.TagKey
import net.minecraft.text.Text

open class AdornReiClient : REIClientPlugin {
    override fun registerCategories(registry: CategoryRegistry) {
        registry.add(BrewerCategory())
        registry.addWorkstations(AdornReiServer.BREWER, EntryStacks.of(AdornBlocks.BREWER))
    }

    override fun registerDisplays(registry: DisplayRegistry) {
        registry.registerRecipeFiller(ItemBrewingRecipe::class.java, AdornRecipes.BREWING_TYPE, ::BrewerDisplay)
        registry.registerRecipeFiller(FluidBrewingRecipe::class.java, AdornRecipes.BREWING_TYPE, ::BrewerDisplay)
    }

    override fun registerCollapsibleEntries(registry: CollapsibleEntryRegistry) {
        registry.add(AdornTags.SOFAS.item)
        registry.add(AdornTags.CHAIRS.item)
        registry.add(AdornTags.TABLES.item)
        registry.add(AdornTags.DRAWERS.item)
        registry.add(AdornTags.KITCHEN_COUNTERS.item)
        registry.add(AdornTags.KITCHEN_CUPBOARDS.item)
        registry.add(AdornTags.KITCHEN_SINKS.item)
        registry.add(AdornTags.POSTS.item)
        registry.add(AdornTags.PLATFORMS.item)
        registry.add(AdornTags.STEPS.item)
        registry.add(AdornTags.SHELVES.item)
        registry.add(AdornTags.CHIMNEYS.item)
        registry.add(AdornTags.COFFEE_TABLES.item)
        registry.add(AdornTags.BENCHES.item)
        registry.add(AdornTags.CRATES.item)
        registry.add(AdornTags.TABLE_LAMPS.item)
        registry.add(AdornTags.CANDLELIT_LANTERNS.item)
    }

    override fun registerScreens(registry: ScreenRegistry) {
        registry.registerDraggableStackVisitor(TradingStationDraggableStackVisitor)
    }

    private fun CollapsibleEntryRegistry.add(tag: TagKey<Item>) {
        val id = tag.id
        val name = Text.translatable("tag.${id.namespace}.${id.path}") // matches the translation keys used by EMI as well
        group(id, name, EntryIngredients.ofItemTag(tag))
    }
}
